### PR TITLE
H-2246: Add Google Sheets integration

### PR DESCRIPTION
**Pull request by CiaranMn**
## 🌟 What is the purpose of this PR?

This PR adds the ability to link a Google account, and write the results of a query for entities to a spreadsheet.

A sheet is created per entity type, and the entities of that type listed as the rows.

It can be formatted for humans, in which case source entities link to the rows of links they are the source of in link type sheets, and the link type sheets link to the row for the source and target of the link. Or it can be formatted for machines, in which case it's plain text with ids for sheet names / column headers instead of titles.

This doesn't yet allow specifying a schedule on which to update the sheet – that will be part of wider "Flows" work. Writing a set of entities to a spreadsheet is one potential node in a Flow graph, and we should discuss how scheduling and error handling for Flows will work generally rather than doing something bespoke here.

## 🔍 What does this change?

- Add migration for Google account and sheets integration entity types
- Add Google OAuth callback handler and token fetching stuff
- Consolidates / standardizes how we create user secrets, since we now have two places we do (for Linear and Google). e.g. make it so that only the bots in question can edit the secret, and the user can view them – no other permissions on either the secret or the link to it. 
- Add functions to the integration Temporal worker to write to a Google sheet
- Add unstyled section to `/settings/integration` to allow for testing out creating and editing jobs to write to a Google sheet

**Drive-bys**
- A couple of tweaks to the HashGPT README and schema.
- Adjust auth rate limiter so it's less likely to catch legitimate OAuth attempts (since they can involve a few visits to Kratos/Hydra)

### 🚢 Has this modified a publishable library?

This PR:

- [x] does not modify any publishable blocks or libraries, or modifications do not need publishing

### 📜 Does this require a change to the docs?

The changes in this PR:

- [x] are in a state where docs changes are not _yet_ required but will be

### 🕸️ Does this require a change to the Turbo Graph?

The changes in this PR:

- [x] do not affect the execution graph

## ⚠️ Known issues

See next steps.

## 🐾 Next steps

Some further improvements tracked in H-2350:
- Expand Terraform config to include Google environment variables, update README to mention them (blocks making this work in production)
- Allow specifying query depth – for this we should expand/refactor the query editor to create structural queries rather than the more limited Block Protocol queries
- Prevent multiple integrations being created that target the same spreadsheet

Other improvements to be wrapped up in wider "Flows" work:
- allowing updating sheets on a schedule
- notify users when a sync fails, e.g. because their credentials have been revoked / otherwise become invalid

## 🛡 What tests cover this?

- None yet.

## ❓ How to test this?

1. Set `GOOGLE_OAUTH_CLIENT_ID` and `GOOGLE_OAUTH_CLIENT_SECRET` from 1Password (look for Google OAuth)
2. You'll need to rebuild the integration Temporal worker or run it locally
3. Go to settings / integrations and link a Google account
4. Choose an existing spreadsheet or name a new one
5. Create a query and click sync

## 📹 Demo

https://github.com/hashintel/hash/assets/37743469/0890ba22-28bd-478a-a96d-a9789d5adea0

### CODE REVIEW

File: apps/hash-integration-worker/src/main.ts
Comment (benwerner01) [inline apps/hash-integration-worker/src/main.ts:75]: **Blocking:** This works when running the integration worker locally, but it doesn't work when running the integration worker in docker. I assume this is because we aren't passing the correct vault environment variables to the container.

File: apps/hash-api/src/integrations/google/oauth-callback.ts
Comment (benwerner01) [inline apps/hash-api/src/integrations/google/oauth-callback.ts:76]: **Blocking:** the `name` field can be `undefined` for some google accounts (e.g. for my `bw@hash.ai` google account). We could account for this by:
- using the email address as the display name, if no `name` is available from google
- making the `displayName` property optional on the google account entity

**Review by benwerner01 — CHANGES_REQUESTED** [review]
Thanks Ciaran, I encountered two issues when running this locally (marked these as **blocking**), but apart from those this the changes look good to me 👍

File: apps/hash-api/src/integrations/google/create-or-update-sheets-integration.ts
Comment (CiaranMn) [reply apps/hash-api/src/integrations/google/create-or-update-sheets-integration.ts:181]: The presence of either `existingSpreadsheetId` or `newFileName` has been checked a few lines above

File: apps/hash-api/src/integrations/google/shared/get-or-check-access-token.ts
Comment (CiaranMn) [reply apps/hash-api/src/integrations/google/shared/get-or-check-access-token.ts:37]: I'm not sure it's worth it, because we'd have to do one of:
1. Pass in the `res` to the function so that it can send the appropriate response if something is missing, which makes it harder to follow the response handling since part of it is buried in another function (I do this in `get-or-check-access-token` and I'm not a big fan of it there either, but that's saving a lot more repeated code)
2. When we get the `user` and `vaultClient` back, in this function send a response if one is missing, in which case we'd have the same amount of repeated code here

File: apps/hash-integration-worker/src/graph-activities.ts
Comment (CiaranMn) [reply apps/hash-integration-worker/src/graph-activities.ts]: Done in https://github.com/hashintel/hash/pull/4141/commits/1ba61d25e65156c55eced5123999df886414dda9

File: apps/hash-frontend/src/pages/settings/integrations/google-sheets/create-or-edit-sheets-integration.tsx
Comment (CiaranMn) [reply apps/hash-frontend/src/pages/settings/integrations/google-sheets/create-or-edit-sheets-integration.tsx:145]: True and fair, but I'm going to leave it for now because we'll probably end up defining these integrations as part of a Flow rather than on this screen – will bear it in mind for when we have the proper designs

File: apps/hash-frontend/src/pages/settings/integrations/google-sheets/create-or-edit-sheets-integration.tsx
Comment (CiaranMn) [reply apps/hash-frontend/src/pages/settings/integrations/google-sheets/create-or-edit-sheets-integration.tsx:236]: Yes, basically. This isn't a great solution, we'd be better off having some proper typed solution to REST endpoints, or just using a GQL resolver. Maybe we can talk about this in the context of flows since creating this integration will probably become creating a Flow that ends up writing entities to a spreadsheet.

File: apps/hash-integration-worker/src/google-activities/convert-subgraph-to-sheet-requests.ts
Comment (CiaranMn) [reply apps/hash-integration-worker/src/google-activities/convert-subgraph-to-sheet-requests.ts]: Thank you, done in https://github.com/hashintel/hash/pull/4141/commits/171223b0f68c2573a2aa1c27a1a5644c0eb88a46

File: apps/hash-api/src/integrations/google/oauth-callback.ts
Comment (CiaranMn) [reply apps/hash-api/src/integrations/google/oauth-callback.ts]: Thank you – I've gone for defaulting to the email address, in https://github.com/hashintel/hash/pull/4141/commits/384f696db89ce4fe68dd0483470dc30e4670d49a

File: apps/hash-api/src/integrations/google/create-or-update-sheets-integration.ts
Comment (benwerner01) [reply apps/hash-api/src/integrations/google/create-or-update-sheets-integration.ts:181]: Ah sorry, missed this

File: apps/hash-api/src/integrations/google/shared/get-or-check-access-token.ts
Comment (benwerner01) [reply apps/hash-api/src/integrations/google/shared/get-or-check-access-token.ts:37]: Alright fair enough, happy to keep as is then

File: apps/hashdotdev/sitemap.json
Comment (benwerner01) [inline apps/hashdotdev/sitemap.json:1]: This file shouldn't be committed, once you merge in changes from `main` it will be in the `.gitignore`.

**Comment by benwerner01** [comment]
Thanks Ciaran, the changes are now working for me locally as expected. Happy to approve once the accidental committed file is removed, and the merge conflicts are resolved 👍

File: apps/hashdotdev/sitemap.json
Comment (CiaranMn) [reply apps/hashdotdev/sitemap.json:1]: Thank you, removed

**Review by benwerner01 — APPROVED** [review]

**Pull request merged**
